### PR TITLE
[plugins.telesync] auto-restrict new users

### PR DESCRIPTION
**Pull request by deffnull**
- can be controlled with per conversation config key “sync_restrict_user”, default off

### CODE REVIEW

File: hangupsbot/sync/__init__.py [hunk 1]
```diff
@@ -131,6 +131,10 @@
     'sync_format_reply_empty': '| <i><b>{name}</b>: {image_tag}~</i>\n',
     'sync_format_reply_bot': '| <i>{image_tag}{text}</i>\n',
     'sync_format_reply_bot_empty': '| <i>{image_tag}~</i>\n',
+
+    # automatically restrict TG users on join, possible values are: messages,
+    # media, sticker, websites, sticker+websites
+    'sync_restrict_user': '',
```
Comment (das7pad) [inline hangupsbot/sync/__init__.py]: Let's keep the telesync defaults in `plugins.telesync.setup_config.default_config` and access them with `bot.get_config_suboption`.

File: hangupsbot/plugins/telesync/core.py [hunk 2]
```diff
@@ -557,6 +566,34 @@ def _on_membership_change(self, msg):
...
+                                                 'telesync:' + msg.chat_id,
+                                                 'restrict_user')
+                rights = (NO_SENDING_RIGHTS if restrict == 'messages' else
+                          NO_MEDIA_RIGHTS if restrict == 'media' else
+                          NO_STICKER_RIGHTS if restrict == 'sticker' else
+                          NO_WEBPREVIEW_RIGHTS if restrict == 'websites' else
+                          NO_WEBPREVIEW_AND_STICKER_RIGHTS if restrict == 'sticker+websites'
+                          else None)
+                if rights and msg['chat']['type'] == 'supergroup':
+                    result = await self.restrictChatMember(msg.chat_id,
```
Comment (das7pad) [inline hangupsbot/plugins/telesync/core.py]: We are actually not in a coroutine here, this requires `async def _on_membership_change(...):` and `await self._on_membership_change(msg)` in line 485.

File: hangupsbot/plugins/telesync/core.py [hunk 3]
```diff
@@ -557,6 +566,34 @@ def _on_membership_change(self, msg):
...
+                                                           changed_member.usr_id,
+                                                           **rights)
+
+                    if result is not True:
+                        mod_chat = self.config('mod_chat') if 'mod_chat' in self.config() else None
```
Comment (das7pad) [inline hangupsbot/plugins/telesync/core.py]: What about an entry in `plugins.telesync.setup_config.default_config` for `mod_chat`?

File: hangupsbot/plugins/telesync/core.py (hunk 2 above)
Comment (das7pad) [reply hangupsbot/plugins/telesync/core.py]: `TelegramBot.restrictChatMember` likely raises an Exception - the membership change would never reach the sync handler in this case.

**Review by das7pad — CHANGES_REQUESTED** [review]
Great idea to add another mod function to `telesync`!

There is a lot of duplicate code here: what about splitting `plugins.telesync.commands_tg.command_restrict_user` into a core and command head:
- the head would enforce an admin user and `args`/valid option and sends the result back
- the core is feed with the `restrict` mode and `args` to parse the specified user_ids; it could simply return the `failed` dict.

We could use the restrict_user core then to run it with the user ids of the changed member.

**Comment by das7pad** [comment]
> what about splitting plugins.telesync.commands_tg.command_restrict_user into a core and command head

done in https://github.com/das7pad/hangoutsbot/commit/71b6ca0d7e2eb7d8af25c4223773a0936eb16677

**Comment by deffnull** [comment]
Changes applied based on 71b6ca0

**Comment by das7pad** [comment]
Thank you!

**Pull request merged**